### PR TITLE
Add SignInManager

### PR DESCRIPTION
**Pull request by neoscie**

### CODE REVIEW

File: Neolution.Extensions.Identity.Abstractions/IJwtSignInManager.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        /// <returns>The <see cref="JsonWebToken"/>.</returns>
+        Task<JsonWebToken?> SignInWithGoogleAsync(string token, string clientId);
+
+        /// <summary>
+        /// Signs in via password.
+        /// </summary>
+        /// <param name="user">The user.</param>
+        /// <param name="password">The password.</param>
+        /// <returns>The <see cref="JsonWebToken"/>.</returns>
+        Task<JsonWebToken?> PasswordSignInAsync(TUser user, string password);
```
Comment (neotrow) [inline Neolution.Extensions.Identity.Abstractions/IJwtSignInManager.cs]: I suggest to also add an overload where we can pass the email. That's probably 99% of time the case.

File: Neolution.Extensions.Identity/JwtSignInManager.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+            // TODO: Think about different discovery options for external users
+            var user = await this.userManager.FindByEmailAsync(payload.Email);
+            if (user is null)
+            {
+                return null;
+            }
+
+            var canSignInResult = await this.signInManager.PreSignInCheckAsync(user);
+            if (canSignInResult != null)
```
Comment (neotrow) [inline Neolution.Extensions.Identity/JwtSignInManager.cs]: This check should check if the canSignInResult is null or ifs not succeeded.

File: Neolution.Extensions.Identity/JwtSignInManager.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            var payload = await GoogleJsonWebSignature.ValidateAsync(token, validationSettings); // Will throw an exception if validation fails.
+
+            // TODO: Think about different discovery options for external users
+            var user = await this.userManager.FindByEmailAsync(payload.Email);
+            if (user is null)
+            {
+                return null;
+            }
+
+            var canSignInResult = await this.signInManager.PreSignInCheckAsync(user);
```
Comment (neotrow) [inline Neolution.Extensions.Identity/JwtSignInManager.cs]: rename it to preSignInResult. Because canSignIn is only part of the check in the Microsoft.Identity world.

File: Neolution.Extensions.Identity/JwtSignInManager.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        /// <inheritdoc/>
+        public async Task<JsonWebToken?> PasswordSignInAsync(TUser user, string password)
+        {
+            // TODO: Decide on making this configurable
+            const bool lockoutOnFailure = true;
+
+            var signInResult = await this.signInManager.CheckPasswordSignInAsync(user, password, lockoutOnFailure);
+            if (!signInResult.Succeeded)
```
Comment (neotrow) [inline Neolution.Extensions.Identity/JwtSignInManager.cs]: can the signInResult also be null?

File: Neolution.Extensions.Identity/JwtSignInManager.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        /// </summary>
+        private readonly IJwtGenerator<TUser> jwtGenerator;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="JwtSignInManager{TUser}"/> class.
+        /// </summary>
+        /// <param name="userManager">The user manager.</param>
+        /// <param name="signInManager">The sign in manager.</param>
+        /// <param name="jwtGenerator">The JWT generator.</param>
+        public JwtSignInManager(IUserManager<TUser> userManager, ISignInManager<TUser> signInManager, IJwtGenerator<TUser> jwtGenerator)
```
Comment (neotrow) [inline Neolution.Extensions.Identity/JwtSignInManager.cs]: Should we call `AccessFailedAsync` somewhere in this method in case the login fails. Or does it not make any sense to support lockout for external providers?

**Review by neotrow — CHANGES_REQUESTED** [review]

File: Neolution.Extensions.Identity.Abstractions/IJwtSignInManager.cs (hunk 1 above)
Comment (neoscie) [reply Neolution.Extensions.Identity.Abstractions/IJwtSignInManager.cs]: TokenSignInManager now takes just username/email and password and returns User if he can be signed in

File: Neolution.Extensions.Identity/JwtSignInManager.cs (hunk 2 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/JwtSignInManager.cs]: Fixed

File: Neolution.Extensions.Identity/JwtSignInManager.cs (hunk 3 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/JwtSignInManager.cs]: fixed

File: Neolution.Extensions.Identity/JwtSignInManager.cs (hunk 4 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/JwtSignInManager.cs]: No

File: Neolution.Extensions.Identity/JwtSignInManager.cs (hunk 5 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/JwtSignInManager.cs]: As discussed, we leave it out for the moment

File: Neolution.Extensions.Identity/Properties/launchSettings.json [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (neotrow) [inline Neolution.Extensions.Identity/Properties/launchSettings.json]: @neoscie 

is this file needed?

File: Neolution.Extensions.Identity/TokenSignManager.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            this.logger = logger;
+            this.userManager = userManager;
+            this.signInManager = signInManager;
+            this.jwtGenerator = jwtGenerator;
+            this.options = options.Value;
+        }
+
+        /// <inheritdoc />
+        public async Task<TUser?> PasswordSignInAsync(string username, string password)
```
Comment (neotrow) [inline Neolution.Extensions.Identity/TokenSignManager.cs]: @neoscie 

there is an inconsitency between username and email. 

For us it's the same but it might not always be the case. 

So either call the parameter `email` or treat it as a `username` by calling `FindByUsernameAsync` and change the logging string

**Review by neotrow — CHANGES_REQUESTED** [review]

File: Neolution.Extensions.Identity/Properties/launchSettings.json (hunk 6 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/Properties/launchSettings.json]: removed

File: Neolution.Extensions.Identity/TokenSignManager.cs (hunk 7 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/TokenSignManager.cs]: fixed

**Review by neotrow — APPROVED** [review]

File: Neolution.Extensions.Identity/TokenSignManager.cs [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+                {
+                    return null;
+                }
+
+                this.logger.LogWarning("Google ID token sign-in for user with id={UserId} failed", user.Id);
+                return user;
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogError(ex, "Could not sign in user with Google ID token");
```
Comment (neotrow) [inline Neolution.Extensions.Identity/TokenSignManager.cs]: @neoscie 

not sure if this should be logged as an error because `GoogleJsonWebSignature.ValidateAsync` will throw an error if an invalid token (expired, wrong client id etc) is passed.

Or maybe add another try catch around the `GoogleJsonWebSignature.ValidateAsync` call.

File: Neolution.Extensions.Identity/TokenSignManager.cs [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            {
+                return await this.CreateAccessTokenAsync(user, null);
+            }
+
+            this.logger.LogWarning("Password sign-in for user email={User} failed", email);
+            return null;
+        }
+
+        /// <inheritdoc />
+        public async Task<JsonWebToken?> TwoFactorAuthenticatorSignInAsync(Guid userId, string code, string? authenticationMethod)
```
Comment (neotrow) [inline Neolution.Extensions.Identity/TokenSignManager.cs:73]: @neoscie 

what is authentication method used for?

**Review by neotrow — CHANGES_REQUESTED** [review]

File: Neolution.Extensions.Identity/TokenSignManager.cs (hunk 9 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/TokenSignManager.cs:73]: It's not (yet) in use. 
The idea is, that if we (ever) sign a JWT for the user, based on a third-party provider like `google` or `microsoft` etc., we should add that information into the JWT. The Cookie-based SignInManager writes this information into the 2fa-cookie and after successful sign-in into the auth-cookie.

File: Neolution.Extensions.Identity/TokenSignManager.cs (hunk 8 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/TokenSignManager.cs]: fixed

File: Neolution.Extensions.Identity.Abstractions/IJwtGenerator.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        /// <summary>
+        /// Generates the access token.
+        /// </summary>
+        /// <param name="user">The user account.</param>
+        /// <param name="userClaims">The user claims.</param>
+        /// <param name="amr">The Authentication Method Reference value.</param>
+        /// <returns>
+        /// The JSON web token
+        /// </returns>
+        JsonWebToken GenerateAccessToken(TUser user, IEnumerable<Claim> userClaims, string amr);
```
Comment (neotrow) [inline Neolution.Extensions.Identity.Abstractions/IJwtGenerator.cs:33]: @neoscie 

couldn't this just be passed as a Claim?

File: Neolution.Extensions.Identity/TokenSignManager.cs [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+            if (authenticationMethod != null)
+            {
+                claims.Add(new Claim(ClaimTypes.AuthenticationMethod, authenticationMethod));
+            }
+
+            return this.jwtGenerator.GenerateAccessToken(user, claims, "mfa");
+        }
+
+        /// <inheritdoc />
+        public async Task<TUser?> GoogleSignInAsync(string token)
```
Comment (neotrow) [inline Neolution.Extensions.Identity/TokenSignManager.cs]: @neoscie do we still need this? 

I think we should remove this and use the `OpenIdConnectSignInAsync` instead.

File: Neolution.Extensions.Identity/TokenSignManager.cs [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+        }
+
+        /// <summary>
+        /// Validates the OpenID connect token. Throws an exception if the token is invalid.
+        /// </summary>
+        /// <param name="token">The token.</param>
+        /// <returns>The <see cref="ClaimsPrincipal"/>.</returns>
+        private async Task<ClaimsPrincipal> ValidateOpenIdConnectInfoAsync(OpenIdConnectToken token)
+        {
+            var configurationManager = new ConfigurationManager<OpenIdConnectConfiguration>(token.DiscoveryDocumentUrl, new OpenIdConnectConfigurationRetriever(), new HttpDocumentRetriever());
```
Comment (neotrow) [inline Neolution.Extensions.Identity/TokenSignManager.cs]: @neoscie  is this going to be cached or downloaded every time?

**Review by neotrow — CHANGES_REQUESTED** [review]

File: Neolution.Extensions.Identity/TokenSignManager.cs (hunk 12 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/TokenSignManager.cs]: There is some sort of cache. I'm not sure about the defaults and how it works exactly. Here is the options class to configure it: https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.configuration.lastknowngoodconfigurationcacheoptions?view=msal-web-dotnet-latest

File: Neolution.Extensions.Identity/TokenSignManager.cs (hunk 11 above)
Comment (neoscie) [reply Neolution.Extensions.Identity/TokenSignManager.cs]: removed

**Review by neotrow — APPROVED** [review]
